### PR TITLE
RolesLookup library

### PR DESCRIPTION
**Pull request by pdyraga**
See https://github.com/keep-network/keep-ecdsa/pull/470

There are four valid cases for unbonded value withdrawal:
- operator withdraws,
- token owner withdraws (when the delegated stake are tokens owned by the delegating party),
- grantee withdraws (when the delegated stake are tokens granted to the delegating party),
- managed grantee withdraws (when the delegated stake are tokens granted to the delegating party via `ManagedGrant` contract).

To detect whether the party is eligible to withdraw, `KeepBonding` needs to resolve the roles and it requires the knowledge about how `TokenStaking` and `TokenGrant` represent relationships. To avoid leaking this logic into `KeepBonding` contract, here we introduce `RolesLookup` library allowing implementing the required checks.

I have also considered implementing a separate contract but:
- The approach with the library and `internal` functions is the least invasive to the current deployment. All we need to do is to pass `TokenGrant` reference to `KeepBonding` contract and `KeepBonding` contract has to be aware of grant relationships anyway to perform ACL checks. All library functions are `internal`, we don't need to deploy this library separately. Changes are minimal and further development should be trivial.
- The separate contract should contain more than just those three functions. Importantly, it should implement all the functions allowing to speed up all dApp roles lookup. I decided to not implement this contract in a hurry, forgetting about half of functionalities it should have and adding incomplete building block on our on-chain components dependency map but rather keep it all in the library and implement new things there when needed. 
- Separate contract is still a good option, I just think we need more time to do it right and from `KeepBonding` perspective, I believe the library with internal functions is better. `KeepBonding` needs to talk to `TokenStaking` anyway for 2nd level authorizations so the separate contract would not reduce the number of dependencies there.

### CODE REVIEW

**Comment by pdyraga** [comment]
I have no idea why is it failing. Executed tests locally multiple times, they are always passing. Sorry @nkuba, I need to leave this one to you... 😢

File: solidity/contracts/libraries/RolesLookup.sol [hunk 1]
```diff
@@ -0,0 +1,66 @@
+pragma solidity 0.5.17;
+
+import "../utils/AddressArrayUtils.sol";
+import "../TokenStaking.sol";
+import "../TokenGrant.sol";
+
+interface ManagedGrant {
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/contracts/libraries/RolesLookup.sol]: https://github.com/keep-network/keep-core/pull/1834/commits/7ff9fed5a63469d6915d83fb3915968d77ef7ccd

[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/contracts/libraries/RolesLookup.sol]: The change fixed the failling tests problem, mentioned in https://github.com/keep-network/keep-core/pull/1834#issuecomment-636382791:


> I have no idea why is it failing. Executed tests locally multiple times, they are always passing. Sorry @nkuba, I need to leave this one to you... 😢

File: solidity/contracts/libraries/RolesLookup.sol [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        TokenGrant tokenGrant
+    ) internal view returns (bool) {
+        address[] memory operators = tokenGrant.getGranteeOperators(grantee);
+        return operators.contains(operator);
+    }
+
+    /// @notice Returns true if the grantee from the given managed granteeContract
+    /// delegated tokens to operator with the provided tokenGrant contract.
+    /// Otherwise, returns false. In case the grantee declared by the managed
+    /// granteeContract does not match the provided grantee, function reverts.
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/contracts/libraries/RolesLookup.sol]: https://github.com/keep-network/keep-core/pull/1834/commits/e510c699d4102ee0e52c31b789bf421667ef5207

File: solidity/contracts/libraries/RolesLookup.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    function isGranteeForOperator(
+        address grantee,
+        address operator,
+        TokenGrant tokenGrant
+    ) internal view returns (bool) {
+        address[] memory operators = tokenGrant.getGranteeOperators(grantee);
+        return operators.contains(operator);
+    }
+
+    /// @notice Returns true if the grantee from the given managed granteeContract
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/contracts/libraries/RolesLookup.sol]: https://github.com/keep-network/keep-core/pull/1834/commits/e510c699d4102ee0e52c31b789bf421667ef5207

File: solidity/contracts/libraries/RolesLookup.sol [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    /// delegated tokens to operator with the provided tokenGrant contract.
+    /// Otherwise, returns false. In case the grantee declared by the managed
+    /// granteeContract does not match the provided grantee, function reverts.
+    /// This function works only for cases when grantee, from TokenGrant's
+    /// perspective, is a smart contract exposing grantee() function returning
+    /// the final grantee. One possibility is the ManagedGrant contract.
+    function isManagedGranteeForOperator(
+        address grantee,
+        address operator,
+        address granteeContract,
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/contracts/libraries/RolesLookup.sol]: https://github.com/keep-network/keep-core/pull/1834/commits/e510c699d4102ee0e52c31b789bf421667ef5207

**Comment by nkuba** [comment]
@pdyraga I pushed some updates to this PR and left two comments that need your opinion.

**Comment by nkuba** [comment]
As the comments I left are non-blocking would you mind doing a final approval/merge @lukasz-zimnoch? We will followup with @pdyraga in a new PR if necessary.

File: solidity/contracts/libraries/RolesLookup.sol [hunk 5]
```diff
@@ -0,0 +1,67 @@
+pragma solidity 0.5.17;
+
+import "../utils/AddressArrayUtils.sol";
+import "../TokenStaking.sol";
+import "../TokenGrant.sol";
+
+
+interface IManagedGrant {
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/libraries/RolesLookup.sol]: I removed this interface in https://github.com/keep-network/keep-core/pull/1834/commits/cc44875993acbef97cc58e4971e519756ab1adcd. After thinking more about it, I think I overcomplicated the code with this interface. This interface was there to allow user-defined grantee contracts work against `RolesLookup`. However, this scenario is not currently supported by `TokenGrant` contract and even if someone wants to implement their own contract-based grantee, they can sill inherit from `ManagedGrant` contract.

**Comment by pdyraga** [comment]
Your changes look good to me @nkuba so if you agree with my recent commit, I think we are good to merge.

File: solidity/contracts/libraries/RolesLookup.sol [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    function grantee() external view returns (address);
+}
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/contracts/libraries/RolesLookup.sol]: Please see https://github.com/keep-network/keep-core/pull/1834#discussion_r435256689

**Pull request merged**